lib/manageiq/providers/openstack/legacy/events/openstack_stf_event_monitor.rb

```diff
@@ -0,0 +1,112 @@
+
+  def self.available?(options = {})
+    $log.info("Testing connection to STF..") if $log
+    $log.debug("With STF options: #{options.inspect}") if $log
+    qdr_client = Qpid::Proton::Container.new(OpenStackStfEventTestReceiver.new(build_qdr_client_url(options), DEFAULT_TOPIC_NAME))
+    qdr_client.run
+    true
+  rescue => ex
+    $log.info("OpenstackSTFEventMonitor availability check failed with #{ex}.") if $log
+    false
+  ensure
+    qdr_client.close if qdr_client.respond_to?(:close)
+    qdr_client.stop if qdr_client.respond_to?(:stop)
+  end
+
+  def self.plugin_priority
+    1
+  end
+
+  def self.build_qdr_client_url(options)
+    $log.debug("Building STF QDR client connection with #{options.inspect}..") if $log
+    protocol = options[:security_protocol].to_s.start_with?('ssl') ? 'amqps' : 'amqp'
+    hostname = options[:hostname]
+    port     = options[:port] || DEFAULT_AMQP_PORT
+    "#{protocol}://#{hostname}:#{port}"
+  end
+
+  def initialize(options = {})
+    $log.info("Building STF QDR client INIT #{options.inspect}..") if $log
+    @options = options
+    @ems = options[:ems]
+    @config = options.fetch(:stf, {})
+
+    @events = []
+    @events_mutex = Mutex.new
+
+    @recv_block = ->(event) { @events << event }
+
+    @qdr_receiver = Qpid::Proton::Container.new(OpenStackStfEventReceiver.new(self.class.build_qdr_client_url(@options), @config[:topic_name] || DEFAULT_TOPIC_NAME, @recv_block, @events_mutex))
+  end
+
+  def start
+    return if @qdr_receiver.running > 0
+
+    $log.info("STF QDR client START..") if $log
+    @handler = Thread.start { @qdr_receiver.run }
+  end
+
+  def stop
+    $log.info("STF QDR client STOP..") if $log
+    @qdr_receiver&.stop
+    @handler&.terminate!
+  end
+
+  def each_batch
+    @collecting_events = true
+    while @collecting_events
+      @events_mutex.synchronize do
+        converted_events = @events.map do |raw_event|
+          unserialized_event = unserialize_event(raw_event)
+          converted_event = OpenstackStfEventConverter.new(unserialized_event)
+          $log.debug("Processing a new OpenStack STF Event: #{unserialized_event.inspect}") if $log
+          openstack_event(nil, converted_event.metadata, converted_event.payload)
+        end
+
+        filtered_events = filter_event_types(converted_events)
+
+        $log.info("MIQ(#{self.class.name}) STF Yielding #{filtered_events.size} events to"\
+        " event_catcher: #{filtered_events.map { |e| e }}") if $log
+
+        yield filtered_events
+
+        $log.info("MIQ(#{self.class.name}) Clearing events") if $log && @events.any? && filtered_events.any?
+        @events.clear
+      end
+      sleep 5
+    end
+  end
+
+  def each
+    each_batch do |events|
+      events.each { |e| yield e }
+    end
+  end
+
+  def unserialize_event(raw_event)
+    msg = JSON.parse(raw_event)
+    event = JSON.parse(msg["request"]["oslo.message"])
+    event["payload"] = event["payload"].first
```

Comment:
Is payload always an array of one?

Reply:
I know this looks strange, the whole process of wrapping internal RabbitMQ Event by Ceilometer backend for distribution via QPID message bus is not well suited for this use case.

QPID/STF Event is generated when a RabbitMQ Event is received, no buffering, no aggregations, it is just wrapped into payload and sent. Confirmed with OSP telemetry team.